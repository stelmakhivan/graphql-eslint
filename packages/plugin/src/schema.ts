import { GraphQLSchema } from 'graphql';
import { GraphQLProjectConfig } from 'graphql-config';
import { asArray } from '@graphql-tools/utils';
import { ParserOptions } from './types';
import { loaderCache } from './utils';

const schemaCache = new Map<string, GraphQLSchema>();

export function getSchema(projectForFile: GraphQLProjectConfig, options: ParserOptions = {}): GraphQLSchema | null {
  const schemaKey = asArray(projectForFile.schema).sort().join(',');
  if (!schemaKey) {
    return null;
  }
  let schema = schemaCache.get(schemaKey);

  if (!schema) {
    schema = projectForFile.loadSchemaSync(projectForFile.schema, 'GraphQLSchema', {
      cache: loaderCache,
      ...options.schemaOptions,
    });
    schemaCache.set(schemaKey, schema);
  }

  return schema;
}
