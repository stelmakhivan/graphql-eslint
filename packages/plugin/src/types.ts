import { Rule, AST, Linter } from 'eslint';
import { ASTNode, GraphQLSchema } from 'graphql';
import { GraphQLParseOptions } from '@graphql-tools/utils';
import { IExtensions, IGraphQLProject, DocumentPointer, SchemaPointer } from 'graphql-config';
import { GraphQLESLintRuleListener } from './testkit';
import { GraphQLESTreeNode } from './estree-parser';
import { SiblingOperations } from './sibling-operations';
import { getReachableTypes, getUsedFields } from './graphql-ast';

export interface ParserOptions {
  schema?: SchemaPointer | Record<string, { headers: Record<string, string> }>;
  documents?: DocumentPointer;
  operations?: DocumentPointer; // legacy
  extensions?: IExtensions;
  include?: string | string[];
  exclude?: string | string[];
  projects?: Record<string, IGraphQLProject>;
  schemaOptions?: Omit<GraphQLParseOptions, 'assumeValidSDL'> & {
    headers: Record<string, string>;
  };
  graphQLParserOptions?: Omit<GraphQLParseOptions, 'noLocation'>;
  skipGraphQLConfig?: boolean;
  filePath?: string;
}

export type ParserServices = {
  hasTypeInfo: boolean;
  schema: GraphQLSchema | null;
  siblingOperations: SiblingOperations;
  reachableTypes: typeof getReachableTypes;
  usedFields: typeof getUsedFields;
};

export type GraphQLESLintParseResult = Linter.ESLintParseResult & {
  services: ParserServices;
};

export type GraphQLESLintRuleContext<Options = any[]> = Omit<
  Rule.RuleContext,
  'parserServices' | 'report' | 'options'
> & {
  options: Options;
  report(
    descriptor: Rule.ReportDescriptorMessage &
      Rule.ReportDescriptorOptions &
      ({ node: GraphQLESTreeNode<ASTNode> } | { loc: AST.SourceLocation | { line: number; column: number } })
  ): void;
  parserServices?: ParserServices;
};

export type CategoryType = 'Schema' | 'Operations';

export type RuleDocsInfo<T> = {
  docs: Omit<Rule.RuleMetaData['docs'], 'category' | 'description'> & {
    category: CategoryType | CategoryType[];
    description: `${string}.`;
    requiresSchema?: true;
    requiresSiblings?: true;
    examples?: {
      title: string;
      code: string;
      usage?: T;
    }[];
    configOptions?:
      | T
      | {
          schema?: T;
          operations?: T;
        };
    graphQLJSRuleName?: string;
    isDisabledForAllConfig?: true
  };
};

export type GraphQLESLintRule<Options = any[], WithTypeInfo extends boolean = false> = {
  create(context: GraphQLESLintRuleContext<Options>): GraphQLESLintRuleListener<WithTypeInfo>;
  meta: Omit<Rule.RuleMetaData, 'docs'> & RuleDocsInfo<Options>;
};

export type ValueOf<T> = T[keyof T];
