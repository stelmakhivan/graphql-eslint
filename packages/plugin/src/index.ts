export { rules } from './rules';
export { processors } from './processors';
export * from './parser';
export * from './types';
export * from './testkit';
export { requireGraphQLSchemaFromContext, requireSiblingsOperations } from './utils';

export const configs = Object.fromEntries(
  ['schema-recommended', 'schema-all', 'operations-recommended', 'operations-all'].map(configName => [
    configName,
    { extends: `./configs/${configName}.json` },
  ])
);
