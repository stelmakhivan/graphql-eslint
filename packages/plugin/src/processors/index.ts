import { createGraphqlProcessor } from './code-files';

const EXTRACTABLE_FILE_EXTENSIONS = ['.js', '.jsx', '.ts', '.tsx', 'graphql'];
const processor = createGraphqlProcessor();

export const processors = Object.fromEntries(EXTRACTABLE_FILE_EXTENSIONS.map(( ext) => [ext, processor]));
