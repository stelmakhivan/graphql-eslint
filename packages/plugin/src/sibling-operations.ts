import { resolve } from 'path';
import {
  FragmentDefinitionNode,
  FragmentSpreadNode,
  Kind,
  OperationDefinitionNode,
  SelectionSetNode,
  visit,
  OperationTypeNode,
} from 'graphql';
import { Source, asArray } from '@graphql-tools/utils';
import { GraphQLProjectConfig } from 'graphql-config';
import { loaderCache } from './utils';

export type FragmentSource = { filePath: string; document: FragmentDefinitionNode };
export type OperationSource = { filePath: string; document: OperationDefinitionNode };

export type SiblingOperations = {
  available: boolean;
  getOperations(): OperationSource[];
  getFragments(): FragmentSource[];
  getFragment(fragmentName: string): FragmentSource[];
  getFragmentByType(typeName: string): FragmentSource[];
  getFragmentsInUse(
    baseOperation: OperationDefinitionNode | FragmentDefinitionNode | SelectionSetNode,
    recursive: boolean
  ): FragmentDefinitionNode[];
  getOperation(operationName: string): OperationSource[];
  getOperationByType(operationType: OperationTypeNode): OperationSource[];
};

const handleVirtualPath = (documents: Source[]): Source[] => {
  const filepathMap: Record<string, number> = Object.create(null);

  return documents.map(source => {
    const { location } = source;
    if (['.gql', '.graphql'].some(extension => location.endsWith(extension))) {
      return source;
    }
    filepathMap[location] ??= -1;
    const index = (filepathMap[location] += 1);
    return {
      ...source,
      location: resolve(location, `${index}_document.graphql`),
    };
  });
};

const operationsCache = new Map<string, Source[]>();
const siblingOperationsCache = new Map<Source[], SiblingOperations>();

const getSiblings = (projectForFile: GraphQLProjectConfig): Source[] => {
  const documentsKey = asArray(projectForFile.documents).sort().join(',');
  if (!documentsKey) {
    return [];
  }
  let siblings = operationsCache.get(documentsKey);

  if (!siblings) {
    const documents = projectForFile.loadDocumentsSync(projectForFile.documents, {
      skipGraphQLImport: true,
      cache: loaderCache,
    });
    siblings = handleVirtualPath(documents);
    operationsCache.set(documentsKey, siblings);
  }

  return siblings;
};

export function getSiblingOperations(projectForFile: GraphQLProjectConfig): SiblingOperations {
  const siblings = getSiblings(projectForFile);

  if (siblings.length === 0) {
    let printed = false;

    const noopWarn = () => {
      if (!printed) {
        // eslint-disable-next-line no-console
        console.warn(
          `getSiblingOperations was called without any operations. Make sure to set "parserOptions.operations" to make this feature available!`
        );
        printed = true;
      }
      return [];
    };

    return {
      available: false,
      getFragments: noopWarn,
      getOperations: noopWarn,
      getFragment: noopWarn,
      getFragmentByType: noopWarn,
      getOperation: noopWarn,
      getOperationByType: noopWarn,
      getFragmentsInUse: noopWarn,
    };
  }

  // Since the siblings array is cached, we can use it as cache key.
  // We should get the same array reference each time we get
  // to this point for the same graphql project
  let siblingOperations = siblingOperationsCache.get(siblings);
  if (!siblingOperations) {
    let fragmentsCache: FragmentSource[] | null = null;

    const getFragments = (): FragmentSource[] => {
      if (fragmentsCache === null) {
        const result: FragmentSource[] = [];

        for (const source of siblings) {
          for (const definition of source.document.definitions || []) {
            if (definition.kind === Kind.FRAGMENT_DEFINITION) {
              result.push({
                filePath: source.location,
                document: definition,
              });
            }
          }
        }
        fragmentsCache = result;
      }
      return fragmentsCache;
    };

    let cachedOperations: OperationSource[] | null = null;

    const getOperations = (): OperationSource[] => {
      if (cachedOperations === null) {
        const result: OperationSource[] = [];

        for (const source of siblings) {
          for (const definition of source.document.definitions || []) {
            if (definition.kind === Kind.OPERATION_DEFINITION) {
              result.push({
                filePath: source.location,
                document: definition,
              });
            }
          }
        }
        cachedOperations = result;
      }
      return cachedOperations;
    };

    const getFragment = (name: string) => getFragments().filter(f => f.document.name?.value === name);

    const collectFragments = (
      selectable: SelectionSetNode | OperationDefinitionNode | FragmentDefinitionNode,
      recursive = true,
      collected = new Map<string, FragmentDefinitionNode>()
    ) => {
      visit(selectable, {
        FragmentSpread(spread: FragmentSpreadNode) {
          const name = spread.name.value;
          const fragmentInfo = getFragment(name);

          if (fragmentInfo.length === 0) {
            // eslint-disable-next-line no-console
            console.warn(
              `Unable to locate fragment named "${name}", please make sure it's loaded using "parserOptions.operations"`
            );
            return;
          }
          const fragment = fragmentInfo[0];
          const alreadyVisited = collected.has(name);

          if (!alreadyVisited) {
            collected.set(name, fragment.document);
            if (recursive) {
              collectFragments(fragment.document, recursive, collected);
            }
          }
        },
      });
      return collected;
    };

    siblingOperations = {
      available: true,
      getFragments,
      getOperations,
      getFragment,
      getFragmentByType: typeName => getFragments().filter(f => f.document.typeCondition?.name?.value === typeName),
      getOperation: name => getOperations().filter(o => o.document.name?.value === name),
      getOperationByType: type => getOperations().filter(o => o.document.operation === type),
      getFragmentsInUse: (selectable, recursive = true) => Array.from(collectFragments(selectable, recursive).values()),
    };

    siblingOperationsCache.set(siblings, siblingOperations);
  }
  return siblingOperations;
}
