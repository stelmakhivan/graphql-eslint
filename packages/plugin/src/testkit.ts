import { readFileSync } from 'fs';
import { resolve } from 'path';
import { RuleTester, AST, Linter, Rule } from 'eslint';
import { ASTKindToNode } from 'graphql';
import { codeFrameColumns } from '@babel/code-frame';
import { GraphQLESTreeNode } from './estree-parser';
import { GraphQLESLintRule, ParserOptions } from './types';

export type GraphQLESLintRuleListener<WithTypeInfo extends boolean = false> = {
  [K in keyof ASTKindToNode]?: (node: GraphQLESTreeNode<ASTKindToNode[K], WithTypeInfo>) => void;
} & Record<string, any>;

export type GraphQLValidTestCase<Options> = Omit<RuleTester.ValidTestCase, 'options' | 'parserOptions'> & {
  name?: string;
  options?: Options;
  parserOptions?: ParserOptions;
};

export type GraphQLInvalidTestCase<T> = GraphQLValidTestCase<T> & {
  errors: number | (RuleTester.TestCaseError | string)[];
  output?: string | null;
};

export class GraphQLRuleTester extends RuleTester {
  config: {
    parser: string;
    parserOptions: ParserOptions;
  };

  constructor(parserOptions: ParserOptions = {}) {
    const config = {
      parser: require.resolve('@graphql-eslint/eslint-plugin'),
      parserOptions: {
        ...parserOptions,
        skipGraphQLConfig: true,
      },
    };
    super(config);
    this.config = config;
  }

  fromMockFile(path: string): string {
    return readFileSync(resolve(__dirname, `../tests/mocks/${path}`), 'utf-8');
  }

  runGraphQLTests<Config>(
    name: string,
    rule: GraphQLESLintRule,
    tests: {
      valid: (string | GraphQLValidTestCase<Config>)[];
      invalid: GraphQLInvalidTestCase<Config>[];
    }
  ): void {
    const ruleTests = Linter.version.startsWith('8')
      ? tests
      : {
          valid: tests.valid.map(test => {
            if (typeof test === 'string') {
              return test;
            }
            const { name, ...testCaseOptions } = test;
            return testCaseOptions;
          }),
          invalid: tests.invalid.map(test => {
            // ESLint 7 throws an error on CI - Unexpected top-level property "name"
            const { name, ...testCaseOptions } = test;
            return testCaseOptions;
          }),
        };

    super.run(name, rule as Rule.RuleModule, ruleTests);

    // Skip snapshot testing if `expect` variable is not defined
    if (typeof expect === 'undefined') {
      return;
    }

    const linter = new Linter();
    linter.defineRule(name, rule as Rule.RuleModule);

    for (const testCase of tests.invalid) {
      const verifyConfig = getVerifyConfig(name, this.config, testCase);
      defineParser(linter, verifyConfig.parser);

      const { code, filename } = testCase;

      const messages = linter.verify(code, verifyConfig, { filename });

      for (const message of messages) {
        if (message.fatal) {
          throw new Error(message.message);
        }

        const messageForSnapshot = visualizeEslintMessage(code, message);
        expect(messageForSnapshot).toMatchSnapshot();
      }
    }
  }
}

function getVerifyConfig(ruleId: string, testerConfig, testCase) {
  const { options, parserOptions, parser = testerConfig.parser } = testCase;

  return {
    ...testerConfig,
    parser,
    parserOptions: {
      ...testerConfig.parserOptions,
      ...parserOptions,
    },
    rules: {
      [ruleId]: options ? ['error', ...options] : 'error',
    },
  };
}

const parsers = new WeakMap<Linter, Set<string>>();

function defineParser(linter: Linter, parser: string): void {
  if (!parser) {
    return;
  }
  if (!parsers.has(linter)) {
    parsers.set(linter, new Set());
  }

  const defined = parsers.get(linter);
  if (!defined.has(parser)) {
    defined.add(parser);
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    linter.defineParser(parser, require(parser));
  }
}

function visualizeEslintMessage(text: string, result: Linter.LintMessage): string {
  const { line, column, endLine, endColumn, message } = result;
  const location: Partial<AST.SourceLocation> = {
    start: {
      line,
      column,
    },
  };

  if (typeof endLine === 'number' && typeof endColumn === 'number') {
    location.end = {
      line: endLine,
      column: endColumn,
    };
  }

  return codeFrameColumns(text, location as AST.SourceLocation, {
    linesAbove: Number.POSITIVE_INFINITY,
    linesBelow: Number.POSITIVE_INFINITY,
    message,
  });
}
