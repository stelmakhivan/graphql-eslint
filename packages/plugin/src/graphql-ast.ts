import { ASTNode, ASTVisitor, TypeInfo, GraphQLSchema, visit, isInterfaceType, visitWithTypeInfo } from 'graphql';
import { SiblingOperations } from './sibling-operations';
import { getTypeName } from './utils';

export type ReachableTypes = Set<string>;

let reachableTypesCache: ReachableTypes;

export function getReachableTypes(schema: GraphQLSchema): ReachableTypes {
  // We don't want cache reachableTypes on test environment
  // Otherwise reachableTypes will be same for all tests
  if (process.env.NODE_ENV !== 'test' && reachableTypesCache) {
    return reachableTypesCache;
  }
  const reachableTypes: ReachableTypes = new Set();

  const collect = (node: ASTNode): void => {
    const typeName = getTypeName(node);
    if (reachableTypes.has(typeName)) {
      return;
    }
    reachableTypes.add(typeName);
    const type = schema.getType(typeName) || schema.getDirective(typeName);

    if (isInterfaceType(type)) {
      const { objects, interfaces } = schema.getImplementations(type);
      for (const { astNode } of [...objects, ...interfaces]) {
        visit(astNode, visitor);
      }
    } else {
      visit(type.astNode, visitor);
    }
  };

  const visitor: ASTVisitor = {
    InterfaceTypeDefinition: collect,
    ObjectTypeDefinition: collect,
    InputValueDefinition: collect,
    UnionTypeDefinition: collect,
    FieldDefinition: collect,
    Directive: collect,
    NamedType: collect,
  };

  for (const type of [
    schema, // visiting SchemaDefinition node
    schema.getQueryType(),
    schema.getMutationType(),
    schema.getSubscriptionType(),
  ]) {
    if (type) {
      visit(type.astNode, visitor);
    }
  }
  reachableTypesCache = reachableTypes;
  return reachableTypesCache;
}

export type UsedFields = Record<string, Set<string>>;

let usedFieldsCache: UsedFields;

export function getUsedFields(schema: GraphQLSchema, operations: SiblingOperations): UsedFields {
  // We don't want cache usedFields on test environment
  // Otherwise usedFields will be same for all tests
  if (process.env.NODE_ENV !== 'test' && usedFieldsCache) {
    return usedFieldsCache;
  }
  const usedFields: UsedFields = Object.create(null);
  const typeInfo = new TypeInfo(schema);

  const visitor = visitWithTypeInfo(typeInfo, {
    Field(node): false | void {
      const fieldDef = typeInfo.getFieldDef();
      if (!fieldDef) {
        // skip visiting this node if field is not defined in schema
        return false;
      }
      const parentTypeName = typeInfo.getParentType().name;
      const fieldName = node.name.value;

      usedFields[parentTypeName] ??= new Set();
      usedFields[parentTypeName].add(fieldName);
    },
  });

  const allDocuments = [...operations.getOperations(), ...operations.getFragments()];
  for (const { document } of allDocuments) {
    visit(document, visitor);
  }
  usedFieldsCache = usedFields;
  return usedFieldsCache;
}
