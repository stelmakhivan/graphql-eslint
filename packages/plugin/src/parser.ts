import { parseGraphQLSDL } from '@graphql-tools/utils';
import { GraphQLError, TypeInfo, Source } from 'graphql';
import { convertToESTree } from './estree-parser';
import { GraphQLESLintParseResult, ParserOptions, ParserServices } from './types';
import { getOnDiskFilepath, extractTokens } from './utils';
import { getSchema } from './schema';
import { getSiblingOperations } from './sibling-operations';
import { loadGraphqlConfig } from './graphql-config';
import { getReachableTypes, getUsedFields } from './graphql-ast';

export function parseForESLint(code: string, options: ParserOptions = {}): GraphQLESLintParseResult {
  const gqlConfig = loadGraphqlConfig(options);
  const filePath = options.filePath || '';
  const realFilepath = filePath ? getOnDiskFilepath(filePath) : null;
  const projectForFile = realFilepath ? gqlConfig.getProjectForFile(realFilepath) : gqlConfig.getDefault();

  const schema = getSchema(projectForFile, options);
  const parserServices: ParserServices = {
    schema,
    hasTypeInfo: schema !== null,
    siblingOperations: getSiblingOperations(projectForFile),
    reachableTypes: getReachableTypes,
    usedFields: getUsedFields,
  };

  try {
    const { document } = parseGraphQLSDL(filePath, code, {
      ...options.graphQLParserOptions,
      noLocation: false,
    });

    const { rootTree, comments } = convertToESTree(document, schema ? new TypeInfo(schema) : null);

    return {
      services: parserServices,
      ast: {
        comments,
        type: 'Program',
        sourceType: 'script',
        body: [rootTree as any],
        loc: rootTree.loc,
        range: rootTree.range as [number, number],
        tokens: extractTokens(new Source(code, filePath)),
      },
    };
  } catch (error) {
    error.message = `[graphql-eslint]: ${error.message}`;
    // In case of GraphQL parser error, we report it to ESLint as a parser error that matches the requirements
    // of ESLint. This will make sure to display it correctly in IDEs and lint results.
    if (error instanceof GraphQLError) {
      const eslintError = {
        index: error.positions[0],
        lineNumber: error.locations[0].line,
        column: error.locations[0].column,
        message: error.message,
      };
      throw eslintError;
    }
    throw error;
  }
}
