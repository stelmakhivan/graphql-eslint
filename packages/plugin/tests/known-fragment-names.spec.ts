import { join } from 'path';
import { GraphQLRuleTester, rules } from '../src';

const ruleTester = new GraphQLRuleTester();

ruleTester.runGraphQLTests('known-fragment-names', rules['known-fragment-names'], {
  valid: [
    {
      filename: join(__dirname, 'mocks/user.graphql'),
      code: ruleTester.fromMockFile('user.graphql'),
      parserOptions: {
        schema: join(__dirname, 'mocks/user-schema.graphql'),
        operations: join(__dirname, 'mocks/user-fields-with-variables.gql'),
      },
    },
    {
      filename: join(__dirname, 'mocks/known-fragment-names.ts/1_document.graphql'),
      code: /* GraphQL */ `
        query User {
          user {
            ...UserFields
          }
        }
      `,
      parserOptions: {
        schema: join(__dirname, 'mocks/user-schema.graphql'),
        operations: join(__dirname, 'mocks/known-fragment-names.ts'),
      },
    },
    {
      name: 'should import all fragments inside fragments',
      filename: join(__dirname, 'mocks/known-fragment-names/user.gql'),
      code: ruleTester.fromMockFile('known-fragment-names/user.gql'),
      parserOptions: {
        schema: join(__dirname, 'mocks/user-schema.graphql'),
        operations: [
          join(__dirname, 'mocks/known-fragment-names/user.gql'),
          join(__dirname, 'mocks/known-fragment-names/user-fields.gql'),
        ],
      },
    },
  ],
  invalid: [],
});
