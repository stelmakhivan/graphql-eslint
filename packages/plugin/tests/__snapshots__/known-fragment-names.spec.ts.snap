// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
  1 | {
  2 |   user {
> 3 |     ...DoesNotExist
    |        ^ Unknown fragment "DoesNotExist".
  4 |   }
  5 | }
  6 |
`;
